From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Wed, 25 Aug 2021 00:56:34 +0200
Subject: adds around screen capturer

This patch adds a new entire screen capturer implementation that is based on CGWindowListCreateImageFromArray to capture the display every frame, thus providing perfect exclusion of around windows.

diff --git a/modules/desktop_capture/BUILD.gn b/modules/desktop_capture/BUILD.gn
index 067287c01e00e91ba0e8b7b0fe264468bba713eb..95ba4955290abbc9ae58840c8064cceb04dcca36 100644
--- a/modules/desktop_capture/BUILD.gn
+++ b/modules/desktop_capture/BUILD.gn
@@ -294,6 +294,7 @@ if (is_mac) {
       "mac/screen_capturer_mac.mm",
       "mouse_cursor_monitor_mac.mm",
       "screen_capturer_darwin.mm",
+      "screen_capturer_mac_around.mm",
       "window_capturer_mac.mm",
       "window_finder_mac.h",
       "window_finder_mac.mm",
diff --git a/modules/desktop_capture/desktop_and_cursor_composer.cc b/modules/desktop_capture/desktop_and_cursor_composer.cc
index fb1afcd37924f6e3b0960e580135e79f88c42900..93cc5610013856b02f722ba5a20d6531d54aa3e3 100644
--- a/modules/desktop_capture/desktop_and_cursor_composer.cc
+++ b/modules/desktop_capture/desktop_and_cursor_composer.cc
@@ -149,14 +149,16 @@ DesktopFrameWithCursor::~DesktopFrameWithCursor() {
 
 DesktopAndCursorComposer::DesktopAndCursorComposer(
     std::unique_ptr<DesktopCapturer> desktop_capturer,
-    const DesktopCaptureOptions& options)
+    const DesktopCaptureOptions& options,
+    bool using_around_capturer)
     : DesktopAndCursorComposer(desktop_capturer.release(),
-                               MouseCursorMonitor::Create(options).release()) {}
+                               MouseCursorMonitor::Create(options).release(), using_around_capturer) {}
 
 DesktopAndCursorComposer::DesktopAndCursorComposer(
     DesktopCapturer* desktop_capturer,
-    MouseCursorMonitor* mouse_monitor)
-    : desktop_capturer_(desktop_capturer), mouse_monitor_(mouse_monitor) {
+    MouseCursorMonitor* mouse_monitor,
+    bool using_around_capturer)
+    : desktop_capturer_(desktop_capturer), mouse_monitor_(mouse_monitor), using_around_capturer_(using_around_capturer) {
   RTC_DCHECK(desktop_capturer_);
 }
 
@@ -166,13 +168,15 @@ std::unique_ptr<DesktopAndCursorComposer>
 DesktopAndCursorComposer::CreateWithoutMouseCursorMonitor(
     std::unique_ptr<DesktopCapturer> desktop_capturer) {
   return std::unique_ptr<DesktopAndCursorComposer>(
-      new DesktopAndCursorComposer(desktop_capturer.release(), nullptr));
+      new DesktopAndCursorComposer(desktop_capturer.release(), nullptr, false));
 }
 
 void DesktopAndCursorComposer::Start(DesktopCapturer::Callback* callback) {
   callback_ = callback;
-  if (mouse_monitor_)
+  if (mouse_monitor_) {
     mouse_monitor_->Init(this, MouseCursorMonitor::SHAPE_AND_POSITION);
+    mouse_monitor_->using_around_capturer_ = using_around_capturer_;
+  }
   desktop_capturer_->should_exclude_ = should_exclude_;
   desktop_capturer_->Start(this);
 }
diff --git a/modules/desktop_capture/desktop_and_cursor_composer.h b/modules/desktop_capture/desktop_and_cursor_composer.h
index a02705810c0011011fbf4ce003bc0f0ae74317c3..f07202ebf3a7edfa273606516b082f6ae664d0dd 100644
--- a/modules/desktop_capture/desktop_and_cursor_composer.h
+++ b/modules/desktop_capture/desktop_and_cursor_composer.h
@@ -37,7 +37,8 @@ class RTC_EXPORT DesktopAndCursorComposer
   // MouseCursorMonitor::Create(options) and renders it into the frames
   // generated by `desktop_capturer`.
   DesktopAndCursorComposer(std::unique_ptr<DesktopCapturer> desktop_capturer,
-                           const DesktopCaptureOptions& options);
+                           const DesktopCaptureOptions& options,
+                           bool using_around_capturer);
 
   ~DesktopAndCursorComposer() override;
 
@@ -69,7 +70,8 @@ class RTC_EXPORT DesktopAndCursorComposer
   // Constructor to delegate both deprecated and new constructors and allows
   // test cases to use a fake MouseCursorMonitor implementation.
   DesktopAndCursorComposer(DesktopCapturer* desktop_capturer,
-                           MouseCursorMonitor* mouse_monitor);
+                           MouseCursorMonitor* mouse_monitor,
+                           bool using_around_capturer);
 
   // DesktopCapturer::Callback interface.
   void OnCaptureResult(DesktopCapturer::Result result,
@@ -84,6 +86,7 @@ class RTC_EXPORT DesktopAndCursorComposer
   DesktopVector cursor_position_;
   DesktopRect previous_cursor_rect_;
   bool cursor_changed_ = false;
+  bool using_around_capturer_ = false;
 
   RTC_DISALLOW_COPY_AND_ASSIGN(DesktopAndCursorComposer);
 };
diff --git a/modules/desktop_capture/mac/desktop_frame_cgimage.h b/modules/desktop_capture/mac/desktop_frame_cgimage.h
index 87804382f23ee913cae65e5f07ca8d54e43aa22c..a76e1b73bb63fe6ca90799d78434d937becffbf3 100644
--- a/modules/desktop_capture/mac/desktop_frame_cgimage.h
+++ b/modules/desktop_capture/mac/desktop_frame_cgimage.h
@@ -33,6 +33,8 @@ class DesktopFrameCGImage final : public DesktopFrame {
   static std::unique_ptr<DesktopFrameCGImage> CreateForWindow(
       CGWindowID window_id);
 
+  static std::unique_ptr<DesktopFrameCGImage> CreateForWindowList(CFArrayRef window_list, CGRect screen_bounds);
+
   ~DesktopFrameCGImage() override;
 
  private:
diff --git a/modules/desktop_capture/mac/desktop_frame_cgimage.mm b/modules/desktop_capture/mac/desktop_frame_cgimage.mm
index fb13fe273880727dd102710dd626a772c4183c51..c94bb272e444d856aeee7337d6312d80500c4b1a 100644
--- a/modules/desktop_capture/mac/desktop_frame_cgimage.mm
+++ b/modules/desktop_capture/mac/desktop_frame_cgimage.mm
@@ -33,7 +33,20 @@
       CGWindowListCreateImage(CGRectNull,
                               kCGWindowListOptionIncludingWindow,
                               window_id,
-                              kCGWindowImageBoundsIgnoreFraming));
+                              kCGWindowImageNominalResolution | kCGWindowImageShouldBeOpaque | kCGWindowImageBoundsIgnoreFraming));
+  if (!cg_image) {
+    return nullptr;
+  }
+
+  return DesktopFrameCGImage::CreateFromCGImage(cg_image);
+}
+
+// static
+std::unique_ptr<DesktopFrameCGImage> DesktopFrameCGImage::CreateForWindowList(CFArrayRef window_list, CGRect screen_bounds) {
+  rtc::ScopedCFTypeRef<CGImageRef> cg_image(
+      CGWindowListCreateImageFromArray(screen_bounds,
+                              window_list,
+                              kCGWindowImageNominalResolution | kCGWindowImageShouldBeOpaque | kCGWindowImageBoundsIgnoreFraming));
   if (!cg_image) {
     return nullptr;
   }
diff --git a/modules/desktop_capture/mouse_cursor_monitor.h b/modules/desktop_capture/mouse_cursor_monitor.h
index e88a5b7201b2cfcb2bb1feb1ad1d8e9d3c11d1dd..507b056065290cb9da89f6d7d355a6f2158daefa 100644
--- a/modules/desktop_capture/mouse_cursor_monitor.h
+++ b/modules/desktop_capture/mouse_cursor_monitor.h
@@ -104,6 +104,8 @@ class MouseCursorMonitor {
   // time) and then Callback::OnMouseCursorPosition() if mode is set to
   // SHAPE_AND_POSITION.
   virtual void Capture() = 0;
+
+  bool using_around_capturer_ = false;
 };
 
 }  // namespace webrtc
diff --git a/modules/desktop_capture/mouse_cursor_monitor_mac.mm b/modules/desktop_capture/mouse_cursor_monitor_mac.mm
index 995aeb9176bc11c72bae3c7cfd200620bc97dedb..bc2ffdab0875ed3b94d34818d01865fa014c1909 100644
--- a/modules/desktop_capture/mouse_cursor_monitor_mac.mm
+++ b/modules/desktop_capture/mouse_cursor_monitor_mac.mm
@@ -114,7 +114,7 @@ void DisplaysReconfigured(CGDirectDisplayID display,
       configuration_monitor_->desktop_configuration();
   float scale = GetScaleFactorAtPosition(configuration, position);
 
-  CaptureImage(scale);
+  CaptureImage(using_around_capturer_ ? 1.0f : scale);
 
   if (mode_ != SHAPE_AND_POSITION)
     return;
diff --git a/modules/desktop_capture/screen_capturer_darwin.mm b/modules/desktop_capture/screen_capturer_darwin.mm
index d5a7bb0522b2d1dbde2f14bf827991ca2d8c4219..6bc3737abeb124dca0256d492a228effe276663f 100644
--- a/modules/desktop_capture/screen_capturer_darwin.mm
+++ b/modules/desktop_capture/screen_capturer_darwin.mm
@@ -11,6 +11,7 @@
 #include <memory>
 
 #include "modules/desktop_capture/mac/screen_capturer_mac.h"
+#include "modules/desktop_capture/screen_capturer_mac_around.h"
 
 namespace webrtc {
 
@@ -21,8 +22,11 @@
     return nullptr;
   }
 
-  std::unique_ptr<ScreenCapturerMac> capturer(new ScreenCapturerMac(
-      options.configuration_monitor(), options.detect_updated_region(), options.allow_iosurface()));
+  // std::unique_ptr<ScreenCapturerMac> capturer(new ScreenCapturerMac(
+  //     options.configuration_monitor(), options.detect_updated_region(), options.allow_iosurface()));
+  std::unique_ptr<ScreenCapturerMacAround> capturer(new ScreenCapturerMacAround(
+      options.configuration_monitor()));
+
   if (!capturer.get()->Init()) {
     return nullptr;
   }
diff --git a/modules/desktop_capture/screen_capturer_mac_around.h b/modules/desktop_capture/screen_capturer_mac_around.h
new file mode 100644
index 0000000000000000000000000000000000000000..f986f397f9c6ce06c7b9315dd6014052c38d0715
--- /dev/null
+++ b/modules/desktop_capture/screen_capturer_mac_around.h
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (c) 2018 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#ifndef MODULES_DESKTOP_CAPTURE_MAC_SCREEN_CAPTURER_MAC_AROUND_H_
+#define MODULES_DESKTOP_CAPTURE_MAC_SCREEN_CAPTURER_MAC_AROUND_H_
+
+#include <ApplicationServices/ApplicationServices.h>
+#include <Cocoa/Cocoa.h>
+#include <CoreGraphics/CoreGraphics.h>
+
+#include <memory>
+#include <vector>
+
+#include "api/scoped_refptr.h"
+#include "modules/desktop_capture/desktop_capturer.h"
+#include "modules/desktop_capture/desktop_capture_options.h"
+#include "modules/desktop_capture/desktop_frame.h"
+#include "modules/desktop_capture/mac/desktop_configuration.h"
+#include "modules/desktop_capture/mac/desktop_configuration_monitor.h"
+#include "modules/desktop_capture/mac/desktop_frame_cgimage.h"
+#include "modules/desktop_capture/mac/window_list_utils.h"
+#include "rtc_base/checks.h"
+#include "rtc_base/constructor_magic.h"
+#include "rtc_base/logging.h"
+#include "rtc_base/trace_event.h"
+
+namespace webrtc {
+
+class ScreenCapturerMacAround : public DesktopCapturer {
+ public:
+  explicit ScreenCapturerMacAround(
+      rtc::scoped_refptr<DesktopConfigurationMonitor> desktop_config_monitor);
+  ~ScreenCapturerMacAround() override;
+
+  bool Init();
+
+  // DesktopCapturer interface.
+  void Start(Callback* callback) override;
+  void CaptureFrame() override;
+  bool GetSourceList(SourceList* sources) override;
+  bool SelectSource(SourceId id) override;
+
+ private:
+  CGRect GetDisplayBounds(CGDirectDisplayID current_display);
+
+  Callback* callback_ = nullptr;
+
+  const rtc::scoped_refptr<DesktopConfigurationMonitor> desktop_config_monitor_;
+
+  // Current display configuration.
+  MacDesktopConfiguration desktop_config_;
+
+  // Currently selected display, or 0 if the full desktop is selected. On OS X
+  // 10.6 and before, this is always 0.
+  CGDirectDisplayID current_display_ = 0;
+  NSRect current_screen_bounds_;
+
+  int64_t last_window_list_collection_ = 0;
+  CFArrayRef window_list_;
+
+  RTC_DISALLOW_COPY_AND_ASSIGN(ScreenCapturerMacAround);
+};
+
+
+}  // namespace webrtc
+
+#endif  // MODULES_DESKTOP_CAPTURE_MAC_SCREEN_CAPTURER_MAC_AROUND_H_
diff --git a/modules/desktop_capture/screen_capturer_mac_around.mm b/modules/desktop_capture/screen_capturer_mac_around.mm
new file mode 100644
index 0000000000000000000000000000000000000000..90819f50204ceb4b1f96c81c6bcb6359fa9324f2
--- /dev/null
+++ b/modules/desktop_capture/screen_capturer_mac_around.mm
@@ -0,0 +1,157 @@
+/*
+ *  Copyright (c) 2013 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#include <utility>
+#include <cmath>
+
+#include "modules/desktop_capture/screen_capturer_mac_around.h"
+
+#include "rtc_base/time_utils.h"
+
+namespace webrtc {
+
+namespace {
+
+CFArrayRef CreateWindowListWithExclusion() {
+  int own_pid = getpid();
+
+  CFArrayRef all_windows =
+      CGWindowListCopyWindowInfo(kCGWindowListOptionOnScreenOnly, kCGNullWindowID);
+  if (!all_windows) return nullptr;
+
+  CFMutableArrayRef returned_array =
+      CFArrayCreateMutable(nullptr, CFArrayGetCount(all_windows), nullptr);
+
+  for (CFIndex i = 0; i < CFArrayGetCount(all_windows); ++i) {
+    CFDictionaryRef window =
+        reinterpret_cast<CFDictionaryRef>(CFArrayGetValueAtIndex(all_windows, i));
+
+    CGWindowID window_id = GetWindowId(window);
+
+    CFNumberRef pid_ref = static_cast<CFNumberRef>(CFDictionaryGetValue(window,
+      kCGWindowOwnerPID));
+    int pid;
+    if (!CFNumberGetValue(pid_ref, kCFNumberIntType, &pid)) continue;
+
+    if (pid != own_pid) {
+      CFArrayAppendValue(returned_array, reinterpret_cast<void*>(window_id));
+    }
+  }
+  CFRelease(all_windows);
+
+  return returned_array;
+}
+
+}
+
+ScreenCapturerMacAround::ScreenCapturerMacAround(
+    rtc::scoped_refptr<DesktopConfigurationMonitor> desktop_config_monitor)
+    : desktop_config_monitor_(std::move(desktop_config_monitor)) {}
+
+ScreenCapturerMacAround::~ScreenCapturerMacAround() {
+  CFRelease(window_list_);
+}
+
+bool ScreenCapturerMacAround::Init() {
+  desktop_config_ = desktop_config_monitor_->desktop_configuration();
+  return true;
+}
+
+bool ScreenCapturerMacAround::GetSourceList(SourceList* screens) {
+  RTC_DCHECK(screens->size() == 0);
+
+  for (MacDisplayConfigurations::iterator it = desktop_config_.displays.begin();
+       it != desktop_config_.displays.end();
+       ++it) {
+    screens->push_back({it->id, std::string()});
+  }
+  return true;
+}
+
+bool ScreenCapturerMacAround::SelectSource(SourceId id) {
+  if (id == kFullDesktopScreenId) {
+    current_display_ = 0;
+  } else {
+    const MacDisplayConfiguration* config =
+        desktop_config_.FindDisplayConfigurationById(static_cast<CGDirectDisplayID>(id));
+    if (!config) return false;
+    current_display_ = config->id;
+  }
+
+  current_screen_bounds_ = GetDisplayBounds(current_display_);
+
+  return true;
+}
+
+void ScreenCapturerMacAround::Start(Callback* callback) {
+  RTC_DCHECK(!callback_);
+  RTC_DCHECK(callback);
+  callback_ = callback;
+
+  window_list_ = CreateWindowListWithExclusion();
+  last_window_list_collection_ = rtc::SystemTimeMillis();
+}
+
+void ScreenCapturerMacAround::CaptureFrame() {
+  TRACE_EVENT0("webrtc", "ScreenCapturerMacAround::CaptureFrame");
+
+  float delta = 1000.0f;
+  bool should_collect_window_list = rtc::TimeSince(last_window_list_collection_) > delta;
+
+  if (should_collect_window_list) {
+    CFRelease(window_list_);
+    window_list_ = CreateWindowListWithExclusion();
+    last_window_list_collection_ = rtc::SystemTimeMillis();
+    current_screen_bounds_ = GetDisplayBounds(current_display_);
+  }
+
+  std::unique_ptr<DesktopFrame> frame = DesktopFrameCGImage::CreateForWindowList(window_list_, current_screen_bounds_);
+  if (!frame) {
+    RTC_LOG(LS_WARNING) << "Temporarily failed to capture window.";
+    callback_->OnCaptureResult(Result::ERROR_TEMPORARY, nullptr);
+    return;
+  }
+
+  DesktopRect bounds = DesktopRect::MakeXYWH(current_screen_bounds_.origin.x, current_screen_bounds_.origin.y,
+    current_screen_bounds_.size.width,
+    current_screen_bounds_.size.height);
+
+  float scale_factor = 1.0f;
+
+  if (!bounds.is_empty() && !frame->size().is_empty()) {
+    float scale_x = frame->size().width() / bounds.width();
+    float scale_y = frame->size().height() / bounds.height();
+    // Currently the scale in X and Y directions must be same.
+    if ((std::fabs(scale_x - scale_y) <=
+         std::numeric_limits<float>::epsilon() * std::max(scale_x, scale_y)) &&
+        scale_x > 0.0f) {
+      scale_factor = scale_x;
+    }
+  }
+
+  frame->mutable_updated_region()->SetRect(
+      DesktopRect::MakeSize(frame->size()));
+  frame->set_top_left(bounds.top_left());
+  frame->set_dpi(DesktopVector(kStandardDPI * scale_factor, kStandardDPI * scale_factor));
+
+  callback_->OnCaptureResult(Result::SUCCESS, std::move(frame));
+}
+
+CGRect ScreenCapturerMacAround::GetDisplayBounds(CGDirectDisplayID current_display) {
+  const MacDisplayConfiguration* config =
+    desktop_config_.FindDisplayConfigurationById(current_display);
+  if (!config) return NSZeroRect;
+  DesktopRect display_bounds = config->bounds;
+
+  return NSMakeRect(display_bounds.left(),
+    display_bounds.top(), display_bounds.width(), display_bounds.height());
+}
+
+}  // namespace webrtc
